### DISCUSSION TITLE
Compress-Archive compared with native Windows 11 GUI "Compress to ZIP file"

### DISCUSSION THREAD

**Discussion by mattmcspirit** [PowerShell/PowerShell#18102] [General]
Hey folks,

I've been experimenting with compressing data on Windows 11 and Windows Server 2022. For testing, I've got a root folder with subfolders, all containing some 10GB VHD files. These are 10GB Fixed size VHDs, but have no actual content inside them. Total is 70GB on disk. I've tested with the following:

1. [IO.Compression.ZipFile]::CreateFromDirectory("V:\ZipTest", "V:\ZipTestNET.zip", "Optimal", $true)
2. Compress-Archive -Path "V:\ZipTest" -DestinationPath "V:\ZipTestPS7.zip" -CompressionLevel SmallestSize (Using PowerShell 7.3.0-preview, and new Archive Module 2.0 Preview 2)
3. Windows 11 GUI -> right-click, Compress to ZIP file

1 and 2 above result in a ~70MB zip file, which itself is a great saving. 3 however, results in a ZIP file that is just **2MB**.

![image](https://user-images.githubusercontent.com/20191274/190485729-3cd8392d-7825-4e1b-b4e7-4daf6f21a0b1.png)

Does anyone have any insight into why the PS-based approaches result in a 35X larger file than the GUI-based compression?

Thanks!

**Comment by jborean93** [comment]
I could be wrong here but I believe the dotnet implementation is something implemented entirely in dotnet itself or at least uses some embeded form of zlib to implement the compression code. This is different from explorer which uses the Shell COM API and is part of Windows with it's own implementation. The separate codebases mean that in some cases you can have slightly different algorithms used to compress the data and in this case it seems like the Windows variant trumps what dotnet/zlib can do. The delta between the 2 is highly dependent on the data that is being compressed and you could find situations where the opposite may happen.

If you want to use PowerShell to create a zip using the Shell/Windows components you can use this code

```powershell
$srcPath = "C:\temp\PowerShell"
$dstPath = "C:\temp\pvpanic.zip"

$shell = New-Object -ComObject Shell.Application
$folder = $shell.NameSpace($srcPath)
$folderItems = $folder.Items()

# Create an empty zip file with the known header that the shell understands
Set-Content -Path $dstPath -Value "PK$([char]5)$([char]6)$(([string][char]0) * 18)" -ErrorAction Stop
$zipTarget = $shell.NameSpace($dstPath)

# https://learn.microsoft.com/en-us/windows/win32/shell/folder-copyhere
$zipTarget.CopyHere($folderItems)

# Call above is async so need a rudimentary check to wait until it's done
while ($folderItems.Count -ne $zipTarget.Items().Count) {
    Start-Sleep -Milliseconds 500
}
```

Unfortunately this has many downsides:

* Requires the shell (explorer) so won't work on Server Core
* You cannot hide the progress window that appears, the standard flags don't seem to work for zip operations here
* The call is async so you need to wait until it's complete using a separate check
  * The above has a rudimentary workaround for this but not sure if this is truly reliable or stable
* Because it uses explorer components to compress the folder it has the same problems as that
  * Non-ASCII chars in filenames will most likely be encoded using the system locale rather than UTF-8 that other tools might expect
  * Unicode filenames won't work due to the above, the system locale won't be able to encode chars that aren't represent in their character set
  * MAX_PATH probably will have problems
* Any interactive prompts will need to be manually handled  unless one of the options for `CopyHere` actually work to override them

Unless you really want to use the Shell variant I highly recommend you use the one in dotnet or a third party implementation that can handle non-interactive scenarios a lot better than what you can do through `Shell.Application`.
